homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,278 @@
+"""Support for interfacing with WS66i 6 zone home audio controller."""
+import logging
+
+from homeassistant import core
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import (
+    CONF_IP_ADDRESS,
+    EVENT_HOMEASSISTANT_STOP,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv, entity_platform, service
+
+from .const import (
+    CONF_SOURCES,
+    DOMAIN,
+    FIRST_RUN,
+    SERVICE_RESTORE,
+    SERVICE_SNAPSHOT,
+    WS66I_OBJECT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_SELECT_SOURCE
+)
+
+
+@core.callback
+def _get_sources_from_dict(data):
+    sources_config = data[CONF_SOURCES]
+
+    source_id_name = {int(index): name for index, name in sources_config.items()}
+
+    source_name_id = {v: k for k, v in source_id_name.items()}
+
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return [source_id_name, source_name_id, source_names]
+
+
+@core.callback
+def _get_sources(config_entry):
+    if CONF_SOURCES in config_entry.options:
+        data = config_entry.options
+    else:
+        data = config_entry.data
+    return _get_sources_from_dict(data)
```

Comment:
Store these in `.options` when you create the config entry so you don't have to check both places